sklearn/_fast_gradient_boosting/loss.pyx

```diff
@@ -0,0 +1,334 @@
+        # Note: a more correct exp is 2 * (raw_predictions - y_true) but
+        # since we use 1 for the constant hessian value (and not 2) this
+        # is strictly equivalent for the leaves values.
+        gradients[i] = raw_predictions[i] - y_true[i]
+
+
+class BinaryCrossEntropy(BaseLoss):
+    """Binary cross-entropy loss, for binary classification.
+
+    For a given sample x_i, the binary cross-entropy loss is defined as the
+    negative log-likelihood of the model which can be expressed as::
+
+        loss(x_i) = log(1 + exp(raw_pred_i)) - y_true_i * raw_pred_i
+
+    See The Elements of Statistical Learning, by Hastie, Tibshirani, Friedman.
+    """
+
+    hessians_are_constant = False
+    inverse_link_function = staticmethod(expit)
+
+    def __call__(self, y_true, raw_predictions, average=True):
+        # shape (1, n_samples) --> (n_samples,). reshape(-1) is more likely to
+        # return a view.
+        raw_predictions = raw_predictions.reshape(-1)
+        # logaddexp(0, x) = log(1 + exp(x))
+        loss = np.logaddexp(0, raw_predictions) - y_true * raw_predictions
+        return loss.mean() if average else loss
+
+    def get_baseline_prediction(self, y_train, prediction_dim):
+        proba_positive_class = np.mean(y_train)
+        eps = np.finfo(y_train.dtype).eps
+        proba_positive_class = np.clip(proba_positive_class, eps, 1 - eps)
+        # log(x / 1 - x) is the anti function of sigmoid, or the link function
+        # of the Binomial model.
+        return np.log(proba_positive_class / (1 - proba_positive_class))
+
+    def update_gradients_and_hessians(self, gradients, hessians, y_true,
+                                      raw_predictions):
+        # shape (1, n_samples) --> (n_samples,). reshape(-1) is more likely to
+        # return a view.
+        raw_predictions = raw_predictions.reshape(-1)
+        gradients = gradients.reshape(-1)
+        hessians = hessians.reshape(-1)
+        return _update_gradients_hessians_binary_crossentropy(
+            gradients, hessians, y_true, raw_predictions)
+
+    def predict_proba(self, raw_predictions):
+        # shape (1, n_samples) --> (n_samples,). reshape(-1) is more likely to
+        # return a view.
+        raw_predictions = raw_predictions.reshape(-1)
+        proba = np.empty((raw_predictions.shape[0], 2), dtype=Y_DTYPE)
+        proba[:, 1] = expit(raw_predictions)
+        proba[:, 0] = 1 - proba[:, 1]
+        return proba
+
+
+cdef void _update_gradients_hessians_binary_crossentropy(
+        G_H_DTYPE_C [::1] gradients,
+        G_H_DTYPE_C [::1] hessians,
+        const Y_DTYPE_C [::1] y_true,
+        const Y_DTYPE_C [::1] raw_predictions):
+    cdef:
+        int n_samples
+        G_H_DTYPE_C gradient_abs
+        int i
+
+    n_samples = raw_predictions.shape[0]
+    for i in prange(n_samples, schedule='static', nogil=True):
+        gradients[i] = cexpit(raw_predictions[i]) - y_true[i]
+        gradient_abs = fabs(gradients[i])
```

Comment:
You're right, but since `y_true[i]` is either 0 or 1 and due to the properties of the sigmoid function (it's in [0, 1]), both are equivalent.

Since lightgbm is doing it this way, I'd rather keep the current version.

Not a proof but might help:

```python
import matplotlib.pyplot as plt
from scipy.special import expit
import numpy as np

x = np.arange(-5, 5, .001)
y = np.abs(expit(x)) * (1 - np.abs(expit(x)))
t = np.abs(expit(x) - 1) * (1 - np.abs(expit(x) - 1))

plt.plot(x, y, label='y')
plt.plot(x, t, label='t')
plt.legend()
plt.show()
```

Reply:
What I still don't understand though is where this absolute value comes from... the derivative of the sigmoid function is `s(x) (1 - s(x))`, without absolute value. If you have any idea I'd be very interested ^^

Reply:
Ha never mind, the absolute value is here to actually **make** the 2 expressions equivalent, and it's applied to the whole gradients not just the sigmoid. I was a bit confused.

Makes me think your solution might be clearer after all, I propose this (same but doesn't use the absolute values since we don't need them):

```python
gradients[i] = cexpit(raw_predictions[i]) - y_true[i]
blahblah = gradients[i] + y_true[i]
hessians[i] = blahblah * (1 - blahblah)  # simple derivative of sigmoid function
```
where I'll find a decent name for `blahblah`

Reply:
I went with

```
        p_i = cexpit(raw_predictions[i])
        gradients[i] = p_i - y_true[i]
        hessians[i] = p_i * (1. - p_i)
```